File: packages/domain/src/booking.ts
import type { ClassSession } from './classes';
import type { Member } from './member';
import type { BusinessRules } from './rules';
import type { TransitionMap } from './shared/machine';
import type { IsoDate } from './shared/time';
import { addHoursIso, isAfter, isBefore, msOf } from './shared/time';

export const BOOKING_STATUSES = [
  'PENDING',
  'CONFIRMED',
  'WAITLIST',
  'CANCELLED',
  'CHECKED_IN',
  'COMPLETED',
  'NO_SHOW',
] as const;
export type BookingStatus = (typeof BOOKING_STATUSES)[number];

export const BOOKING_TRANSITIONS: TransitionMap<BookingStatus> = {
  PENDING: ['CONFIRMED', 'WAITLIST', 'CANCELLED'],
  CONFIRMED: ['CANCELLED', 'CHECKED_IN', 'NO_SHOW'],
  WAITLIST: ['CONFIRMED', 'CANCELLED'],
  CHECKED_IN: ['COMPLETED'],
  CANCELLED: [],
  COMPLETED: [],
  NO_SHOW: [],
};

/** Booking states a member can still act on / that hold a slot or waitlist place. */
export const ACTIVE_BOOKING_STATUSES: readonly BookingStatus[] = [
  'PENDING',
  'CONFIRMED',
  'WAITLIST',
  'CHECKED_IN',
];

export interface Booking {
  id: string;
  memberId: string;
  sessionId: string;
  status: BookingStatus;
  waitlistPosition: number | null;
  source: 'MEMBER' | 'ADMIN';
  createdAt: IsoDate;
  updatedAt: IsoDate;
  cancelledAt: IsoDate | null;
  checkedInAt: IsoDate | null;
  /** Set when a freed slot was offered to this waitlisted member (manual-confirm policy). */
  promotionOfferedAt: IsoDate | null;
}

export type BookingDenialReason =
  | 'MEMBER_NOT_ACTIVE'
  | 'SESSION_NOT_BOOKABLE'
  | 'BOOKING_NOT_OPEN_YET'
  | 'BOOKING_WINDOW_CLOSED'
  | 'ALREADY_BOOKED'
  | 'INSUFFICIENT_CREDITS';

export type BookingDecision =
  | { kind: 'CONFIRM' }
  | { kind: 'WAITLIST'; position: number }
  | { kind: 'DENY'; reason: BookingDenialReason };

/**
 * Credits are deducted at CHECK-IN (gate), not at booking. Booking validates
 * that the member *could* pay, holds the slot, and richer states do the rest.
 */
export function evaluateBookingEligibility(args: {
  member: Member;
  session: ClassSession;
  balance: number;
  confirmedCount: number;
  waitlistCount: number;
  hasExistingActiveBooking: boolean;
  now: IsoDate;
}): BookingDecision {
  const { member, session, now } = args;
  if (member.status !== 'ACTIVE') return { kind: 'DENY', reason: 'MEMBER_NOT_ACTIVE' };
  if (session.status !== 'PUBLISHED' && session.status !== 'FULL')
    return { kind: 'DENY', reason: 'SESSION_NOT_BOOKABLE' };
  if (isBefore(now, session.bookingOpensAt)) return { kind: 'DENY', reason: 'BOOKING_NOT_OPEN_YET' };
  if (isAfter(now, session.bookingClosesAt))
    return { kind: 'DENY', reason: 'BOOKING_WINDOW_CLOSED' };
  if (args.hasExistingActiveBooking) return { kind: 'DENY', reason: 'ALREADY_BOOKED' };
  if (args.balance < session.creditCost) return { kind: 'DENY', reason: 'INSUFFICIENT_CREDITS' };
  if (args.confirmedCount >= session.capacity)
    return { kind: 'WAITLIST', position: args.waitlistCount + 1 };
  return { kind: 'CONFIRM' };
}

export type CancellationOutcome =
  | { kind: 'RELEASED'; deadline: IsoDate }
  | { kind: 'LATE'; deadline: IsoDate; penaltyCredits: number };

export function cancellationDeadline(session: ClassSession, rules: BusinessRules): IsoDate {
  return addHoursIso(session.startsAt, -rules.cancellationDeadlineHours);
}

export function evaluateCancellation(args: {
  session: ClassSession;
  rules: BusinessRules;
  now: IsoDate;
}): CancellationOutcome {
  const deadline = cancellationDeadline(args.session, args.rules);
  if (isAfter(args.now, deadline)) {
    return {
      kind: 'LATE',
      deadline,
      penaltyCredits:
        args.rules.lateCancellationPolicy === 'FORFEIT' ? args.session.creditCost : 0,
    };
  }
  return { kind: 'RELEASED', deadline };
}

export function noShowPenalty(session: ClassSession, rules: BusinessRules): number {
  return rules.noShowPolicy === 'FORFEIT' ? session.creditCost : 0;
}

/** FIFO: lowest waitlist position wins; ties broken by creation time. */
export function pickWaitlistPromotion(waitlist: readonly Booking[]): Booking | null {
  const candidates = waitlist.filter((b) => b.status === 'WAITLIST');
  if (candidates.length === 0) return null;
  return [...candidates].sort(
    (a, b) =>
      (a.waitlistPosition ?? Number.MAX_SAFE_INTEGER) -
        (b.waitlistPosition ?? Number.MAX_SAFE_INTEGER) || msOf(a.createdAt) - msOf(b.createdAt),
  )[0]!;
}

export function nextWaitlistPosition(waitlist: readonly Booking[]): number {
  return (
    waitlist
      .filter((b) => b.status === 'WAITLIST')
      .reduce((max, b) => Math.max(max, b.waitlistPosition ?? 0), 0) + 1
  );
}

File: internal/domain/booking.go
package domain

import (
	"sort"
	"time"
)

// BookingStatus is the lifecycle of one member's place in one session.
type BookingStatus string

const (
	BookingPending   BookingStatus = "PENDING"
	BookingConfirmed BookingStatus = "CONFIRMED"
	BookingWaitlist  BookingStatus = "WAITLIST"
	BookingCancelled BookingStatus = "CANCELLED"
	BookingCheckedIn BookingStatus = "CHECKED_IN"
	BookingCompleted BookingStatus = "COMPLETED"
	BookingNoShow    BookingStatus = "NO_SHOW"
)

var BookingTransitions = TransitionMap[BookingStatus]{
	BookingPending:   {BookingConfirmed, BookingWaitlist, BookingCancelled},
	BookingConfirmed: {BookingCancelled, BookingCheckedIn, BookingNoShow},
	BookingWaitlist:  {BookingConfirmed, BookingCancelled},
	BookingCheckedIn: {BookingCompleted},
	BookingCancelled: {},
	BookingCompleted: {},
	BookingNoShow:    {},
}

// ActiveBookingStatuses are the states that hold a slot or a waitlist place.
// A member may hold only one of these per session at a time.
var ActiveBookingStatuses = []BookingStatus{
	BookingPending, BookingConfirmed, BookingWaitlist, BookingCheckedIn,
}

// IsActive reports whether this booking still occupies a place.
func (b Booking) IsActive() bool { return contains(ActiveBookingStatuses, b.Status) }

// BookingSource records whether the member booked themselves or the front desk
// did it for them.
type BookingSource string

const (
	BookingSourceMember BookingSource = "MEMBER"
	BookingSourceAdmin  BookingSource = "ADMIN"
)

// Booking is a member's place in a session.
type Booking struct {
	ID               string        `json:"id"`
	MemberID         string        `json:"memberId"`
	SessionID        string        `json:"sessionId"`
	Status           BookingStatus `json:"status"`
	WaitlistPosition *int          `json:"waitlistPosition"`
	Source           BookingSource `json:"source"`
	CreatedAt        time.Time     `json:"createdAt"`
	UpdatedAt        time.Time     `json:"updatedAt"`
	CancelledAt      *time.Time    `json:"cancelledAt"`
	CheckedInAt      *time.Time    `json:"checkedInAt"`
	// PromotionOfferedAt is set when a freed slot was offered to this
	// waitlisted member and is waiting for them to confirm.
	PromotionOfferedAt *time.Time `json:"promotionOfferedAt"`
}

// BookingDenialReason explains a refused booking. These strings are API error
// codes: the member app maps each to its own message.
type BookingDenialReason string

const (
	DenyMemberNotActive     BookingDenialReason = "MEMBER_NOT_ACTIVE"
	DenySessionNotBookable  BookingDenialReason = "SESSION_NOT_BOOKABLE"
	DenyBookingNotOpenYet   BookingDenialReason = "BOOKING_NOT_OPEN_YET"
	DenyBookingWindowClosed BookingDenialReason = "BOOKING_WINDOW_CLOSED"
	DenyAlreadyBooked       BookingDenialReason = "ALREADY_BOOKED"
	DenyInsufficientCredits BookingDenialReason = "INSUFFICIENT_CREDITS"
	DenyPackageNotCovered   BookingDenialReason = "PACKAGE_NOT_COVERED"
)

// BookingDecisionKind is the outcome of the eligibility pipeline.
type BookingDecisionKind string

const (
	DecisionConfirm  BookingDecisionKind = "CONFIRM"
	DecisionWaitlist BookingDecisionKind = "WAITLIST"
	DecisionDeny     BookingDecisionKind = "DENY"
)

// BookingDecision carries the outcome plus whichever detail it implies: a
// waitlist position, or the reason for refusal.
type BookingDecision struct {
	Kind     BookingDecisionKind
	Position int
	Reason   BookingDenialReason
}

// BookingRequest is everything the eligibility rules need. Gathering it is the
// caller's job, which keeps this function pure and exhaustively testable.
type BookingRequest struct {
	Member                   Member
	Session                  ClassSession
	Balance                  int
	ConfirmedCount           int
	WaitlistCount            int
	HasExistingActiveBooking bool
	Now                      time.Time
}

// EvaluateBooking decides whether a member may take a slot.
//
// Credits are checked but NOT deducted here: deduction happens at the gate on
// check-in. Booking only proves the member could pay and holds the place.
func EvaluateBooking(req BookingRequest) BookingDecision {
	deny := func(reason BookingDenialReason) BookingDecision {
		return BookingDecision{Kind: DecisionDeny, Reason: reason}
	}
	switch {
	case !req.Member.IsActive():
		return deny(DenyMemberNotActive)
	case !req.Session.IsBookable():
		return deny(DenySessionNotBookable)
	case req.Now.Before(req.Session.BookingOpensAt):
		return deny(DenyBookingNotOpenYet)
	case req.Now.After(req.Session.BookingClosesAt):
		return deny(DenyBookingWindowClosed)
	case req.HasExistingActiveBooking:
		return deny(DenyAlreadyBooked)
	case req.Balance < req.Session.CreditCost:
		return deny(DenyInsufficientCredits)
	case req.ConfirmedCount >= req.Session.Capacity:
		return BookingDecision{Kind: DecisionWaitlist, Position: req.WaitlistCount + 1}
	default:
		return BookingDecision{Kind: DecisionConfirm}
	}
}

// CancellationKind distinguishes a cancellation inside the deadline from one
// after it.
type CancellationKind string

const (
	CancelReleased CancellationKind = "RELEASED"
	CancelLate     CancellationKind = "LATE"
)

// CancellationOutcome reports what the member's cancellation costs them.
type CancellationOutcome struct {
	Kind           CancellationKind
	Deadline       time.Time
	PenaltyCredits int
}

// CancellationDeadline is the last moment a member can cancel for free.
func CancellationDeadline(session ClassSession, rules BusinessRules) time.Time {
	return session.StartsAt.Add(-time.Duration(rules.CancellationDeadlineHours) * time.Hour)
}

// EvaluateCancellation applies the late-cancellation policy.
func EvaluateCancellation(session ClassSession, rules BusinessRules, now time.Time) CancellationOutcome {
	deadline := CancellationDeadline(session, rules)
	if now.After(deadline) {
		penalty := 0
		if rules.LateCancellationPolicy == PolicyForfeit {
			penalty = session.CreditCost
		}
		return CancellationOutcome{Kind: CancelLate, Deadline: deadline, PenaltyCredits: penalty}
	}
	return CancellationOutcome{Kind: CancelReleased, Deadline: deadline}
}

// NoShowPenalty is what a missed class costs under the current policy.
func NoShowPenalty(session ClassSession, rules BusinessRules) int {
	if rules.NoShowPolicy == PolicyForfeit {
		return session.CreditCost
	}
	return 0
}

// PickWaitlistPromotion chooses who gets a freed slot: lowest waitlist
// position first, ties broken by who joined the waitlist earlier.
func PickWaitlistPromotion(waitlist []Booking) *Booking {
	candidates := make([]Booking, 0, len(waitlist))
	for _, b := range waitlist {
		if b.Status == BookingWaitlist {
			candidates = append(candidates, b)
		}
	}
	if len(candidates) == 0 {
		return nil
	}
	sort.SliceStable(candidates, func(i, j int) bool {
		pi, pj := positionOf(candidates[i]), positionOf(candidates[j])
		if pi != pj {
			return pi < pj
		}
		return candidates[i].CreatedAt.Before(candidates[j].CreatedAt)
	})
	winner := candidates[0]
	return &winner
}

// NextWaitlistPosition is the position the next joiner receives.
func NextWaitlistPosition(waitlist []Booking) int {
	highest := 0
	for _, b := range waitlist {
		if b.Status != BookingWaitlist || b.WaitlistPosition == nil {
			continue
		}
		if *b.WaitlistPosition > highest {
			highest = *b.WaitlistPosition
		}
	}
	return highest + 1
}

// positionOf treats an unset position as last in line.
func positionOf(b Booking) int {
	if b.WaitlistPosition == nil {
		return int(^uint(0) >> 1)
	}
	return *b.WaitlistPosition
}
